packages/@functionless/formation/src/graph.ts

```diff
@@ -0,0 +1,254 @@
+  );
+}
+
+export interface TopoEntry {
+  /**
+   * Id of the node
+   */
+  resourceId: string;
+  /**
+   * Depth of the resource, Max depth of all dependencies + 1 or 1
+   */
+  level: number;
+  /**
+   * Unique ordinal where the resource has all dependencies resolved
+   */
+  post: number;
+  /**
+   * Unique ordinal where the resource was first discovered.
+   *
+   * Generally this doesn't provide much value, however the distance between post and pre can help
+   * determine the complexity of the dependencies of the resource.
+   */
+  pre: number;
+}
+
+/**
+ * @param improvedOrder when true, runs the algorithm twice, once with random order keys and again running the lowest
+ *                      most dependency first. This should put dependencies as near to their dependents as possible.
+ */
+export function topoSortWithLevels(
+  graph: DependencyGraph,
+  improvedOrder?: boolean
+): TopoEntry[] {
+  const randomOrderKeys = Object.keys(graph);
+
+  // run once using the random ordered keys
+  const topo1 = topoSort(randomOrderKeys);
+
+  if (!improvedOrder) {
+    return topo1;
+  }
+
+  // run again starting with the lowest depth nodes (highest numbers)
+  // to improve the sorted output order
+  return topoSort(
+    // sort by level descending and run again.
+    topo1
+      .sort((a, b) => (a.post < b.post ? 1 : a.post > b.post ? -1 : 0))
+      .map((a) => a.resourceId)
+  );
+
+  function topoSort(keys: string[]) {
+    let pres: Record<string, number> = {};
+    let posts: Record<string, number> = {};
+    let depth: Record<string, number> = {};
+    let post = 0; // post-order, when all resolvable children are resolved
+    let pre = 0; // pre-order, when the node can be resolved
+
+    console.log(keys);
+
+    keys.forEach(visit);
+
+    return Object.keys(graph)
+      .map((node) => ({
+        resourceId: node,
+        level: depth[node]!,
+        post: posts[node]!,
+        pre: pres[node]!,
+      }))
+      .sort((a, b) => (a.post < b.post ? -1 : a.post > b.post ? 1 : 0));
+
+    function visit(node: string): number {
+      if (node in depth) {
+        // ongoing, circular
+        if (node in pres && !(node in posts)) {
+          throw new Error("Circular reference...");
+        }
+        return depth[node]!;
+      }
+      pres[node] = pre++;
+      const depths = graph[node]!.map(visit);
+      if (depths.length === 0) {
+        depth[node] = 1;
+      } else {
+        depth[node] = Math.max(...depths) + 1;
+      }
+      posts[node] = post++;
+      return depth[node]!;
+    }
+  }
+}
+
+const emptySet = ImmutableSet<string>([]);
+
+function findCircularReferences(graph: DependencyGraph): string[] {
+  const circularReferences: string[] = [];
+  for (const [logicalId, references] of Object.entries(graph)) {
+    for (const reference of references) {
+      const isCircular = transitReference(reference);
+      if (isCircular) {
+        circularReferences.push(logicalId);
+        break;
```

Comment:
Should you continue and find all?

Reply:
I can comment on my own code too. It's all being committed

Reply:
I re-wrote all of this anyways, lol. It was re-doing a lot of work.

Reply:
Can it be removed?